### PR TITLE
Backport isolationtester

### PR DESCRIPTION
**Pull request by hlinnaka**
We desperately need regression tests involving distributed transactions. This PR backports the "isolationtester" from PostgreSQL master, and adds one GPDB test case that uses it, mostly as a demonstration of how to use it. The test case it adds is the repro script from MPP-10963, which is being discussed on gpdb-dev at the moment (https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/JrXnNap84tw/P7_glyDpDgAJ). This PR isn't otherwise related to that discussion, except that having isolationtester available will be very helpful in developing that feature, among other things.

### CODE REVIEW

**Comment by danielgustafsson** [comment]
This is great, :+1:^💯

**Comment by ivannovick** [comment]
+1 and thx for merging this @hlinnaka

**Comment by ashwinstar** [comment]
Thank You Heikki for this backport. Please can you enhance the README to clarify how the blocking is detected for GPDB. Will it detect distributed deadlocks or just works for local locks only right now ?

**Comment by ashwinstar** [comment]
Also, please can you clarify and update REAME
1] Setup doesn't have corresponding out file, what happens if anything fails during setup.
2] Incase of multiple permutations, how outfile is specified ?

**Comment by hlinnaka** [comment]
On 01/06/16 00:14, Ashwin Agrawal wrote:
> Thank You Heikki for this backport. Please can you enhance the README
> to clarify how the blocking is detected for GPDB. Will it detect
> distributed deadlocks or just works for local locks only right now ?

Hmm. It doesn't attempt to detect deadlocks of any sort, only whether a 
session is blocked or not. The README says:

> Note that isolationtester recognizes that a command has blocked by looking
> to see if it is shown as waiting in the pg_locks view; therefore, only
> blocks on heavyweight locks will be detected.

So, it's supposed to detect whatever is visible on pg_locks. But that's 
not actually what the code does: it uses pg_blocking_pids() function, 
not pg_locks view (that text in the README probably ought to be fixed in 
the upstream). They both draw from the proc array. Except that on GPDB, 
selecting from pg_locks also dispatches to the segments, while the new 
pg_blocking_pids(), as I backported it does not.

In summary, as this PR is written, it will only detect blocking on locks 
in the master. Maybe that's OK? Or perhaps we should recant and rip out 
the blocking detection. The first version of isolationtester, in 9.1, 
didn't have that, it was added in 9.2, so it would be quite logical to 
backport the 9.1 version instead.

However, I think we discussed using isolationtester for something that 
would need the support for detecting blocking. I don't remember what it 
was, though. Would you happen to remember?

- Heikki

**Comment by ashwinstar** [comment]
I think the discussion was during distributed deadlock detection to use IsolationTester to run through scenarios. All of our current tests though do have well defined permutations (actually 1) and do not need to rely on framework to detect and get unblocked. The tests are written with the intent to unblock themselves. So, I am on fence but lighter version without blocking check seems okay as start to be less confusing to deal with. Plus, might come with no backend change needed for easy porting to 4.3_STABLE as well :-)

**Comment by ascherbaum-pivotal** [comment]
Looks good. Do you want to merge this before or after the 8.3 merge - and get rid of the 8.3 merge FIXME?

**Comment by hlinnaka** [comment]
On 10/06/16 16:21, Andreas Scherbaum wrote:
> Looks good. Do you want to merge this before or after the 8.3 merge - and get rid of the 8.3 merge FIXME?

Doesn't matter much, I think. But we need to reconsider the "blocking 
detection", per discussion with @ashwinstar. I'm now leaning towards 
ripping that out, essentially having the PostgreSQL 9.1 version of 
isolationtester.

- Heikki

**Comment by hlinnaka** [comment]
Ok, I changed this be a backport the 9,.1 version instead. That doesn't require any backend changes. We can backport the changes from 'master' later, if needed, but this should get us started.

**Comment by hlinnaka** [comment]
Ok, pushed
